### ISSUE TITLE
export formatted references

### ISSUE THREAD

**Issue by panosz** [papis/papis#755]
It would be nice if there was a feature by which a single item can be quickly exported to a csl formatted entry. JabRef for example supports this by <right click> -> copy -> copy citation (text), which gives an output of the form

> [1]S. S. Abdullaev, A. Wingen, and K. H. Spatschek, “Mapping of drift surfaces in toroidal systems with chaotic magnetic fields,” Phys. Plasmas, vol. 13, no. 4, 2006.

This is very useful in a number of situations such as 
1. when writing quick notes which I may need to share and don't want to bother setting up an entire tex project.
2. adding references to docstrings in a python module

and so on
Is there support for such a feature in papis?
If not, can anyone share any useful work arrounds?

**Comment by alexfikl** [comment]
I'm afraid we don't support CSL at the moment. We're mostly held back by a good well-maintained library for Python. There's https://github.com/brechtm/citeproc-py, but hasn't received major updates in quite a while..

I'm not sure what a nice and simple workaround would be. Maybe use `papis list`? This should give something like that
```
papis list --format '[1] {doc[author]}, "{doc[title]}", {doc[journal]}, vol. {doc[volume]}, {doc[year]}'
```
? It's not very robust, but it works..

**Comment by panosz** [comment]
@alexfikl Thank you for your swift reply!
It's a great quick and dirty solution.
Ideally I would like to integrate this to neovim so that I can insert formatted references directly from my editor, but I believe this is a discussion for papis.nvim

**Comment by supersambo** [comment]
Some time ago I played around with pybtex to achieve something like this. The relevant code is  [here](https://github.com/supersambo/papis-tui/blob/main/papistui/features/bibtexprint.py) maybe you can get some helpful snippets out of there.

**Comment by alexfikl** [comment]
> Some time ago I played around with pybtex to achieve something like this. The relevant code is [here](https://github.com/supersambo/papis-tui/blob/main/papistui/features/bibtexprint.py?rgh-link-date=2024-03-01T16%3A26%3A34Z) maybe you can get some helpful snippets out of there.

The issue with that is that `pybtex` doesn't seem to be maintained anymore. I don't think it even works with Python 3.12 :\

**Comment by supersambo** [comment]
> The issue with that is that `pybtex` doesn't seem to be maintained anymore. I don't think it even works with Python 3.12 :\

You are right, sorry. This was just meant to be another hacky/quick approach rather than a lasting solution (This is the reason I did not suggest it elsewhere). I was aware that `pybtex` is not being maintained anymore but I did not know it would not even work with Python 3.12

However, I do think formatting sources in a specific citation style is a crucial feature in a reference manager. Maybe one approach for now could be to export documents as bibtex and then use some latex templates to do the formatting. I imagine this could be done as an addon script for those who need it.

**Comment by alexfikl** [comment]
> However, I do think formatting sources in a specific citation style is a crucial feature in a reference manager.

Completely agree, hooking into all the CSL styles would be quite awesome!

> Maybe one approach for now could be to export documents as bibtex and then use some latex templates to do the formatting. I imagine this could be done as an addon script for those who need it.

I'm mostly a heavy LaTeX user, so exporting to BibTeX and letting the BibTeX engine handle the different styles is my usual workflow. There's also stuff like https://github.com/zepinglee/citeproc-lua.

**Comment by jghauser** [comment]
> I'm mostly a heavy LaTeX user, so exporting to BibTeX and letting the BibTeX engine handle the different styles is my usual workflow. There's also stuff like https://github.com/zepinglee/citeproc-lua.

I've actually been looking at `citeproce-lua` recently as @panosz submitted a [nice PR](https://github.com/jghauser/papis.nvim/pull/33) to `papis.nvim` for implementing a simplified version of this feature. If I can get it to work semi-easily, at least `papis.nvim` might soonish get something like this feature :).

**Comment by alexfikl** [comment]
Some initial CSL support using `citeproc-py` has been added in #976. Unfortunately, from my testing of the library, some common styles (like APA and MLA) don't work yet. There are some PRs that seem to fix that, so hopefully by the time we release a new version, that'll be there too.

In the meantime, I'm going to close this. Feel free to open a new issue if there are any bugs we can fix or improved workflows to add! 😁
